### PR TITLE
Compiler changes related to `on-fail-check` node

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**
Fixes #<Issue Number>

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java [hunk 1]
```diff
@@ -1494,6 +1495,17 @@ public void visit(BLangCheckPanickedExpr source) {
...
 
+    @Override
+    public void visit(BLangCheckedOnFailExpr source) {
+
+        BLangCheckedOnFailExpr clone = new BLangCheckedOnFailExpr();
+        source.cloneRef = clone;
+        clone.bLangCheckedExpr = clone(source.bLangCheckedExpr);
+        clone.bLangIdentifier = clone(source.bLangIdentifier);
+        clone.bLangErrorConstructorExpr = clone(source.bLangErrorConstructorExpr);
+//        clone.isRedundantChecking = source.isRedundantChecking;
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/NodeCloner.java:1506]: if this not needed remove it.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 2]
```diff
@@ -3235,6 +3236,14 @@ public void visit(BLangCheckPanickedExpr checkPanicExpr, AnalyzerData data) {
         analyzeExpr(checkPanicExpr.expr, data);
     }
 
+    @Override
+    public void visit(BLangCheckedOnFailExpr checkedOnFailExpr, AnalyzerData data) {
+        visit(checkedOnFailExpr.getbLangCheckedExpr(), data);
+        visit(checkedOnFailExpr.getbLangIdentifier(), data);
+        visit((BLangErrorConstructorExpr) checkedOnFailExpr.bLangErrorConstructorExpr() , data);
+        // avoid unused variable error
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:3244]: Why this comment in the end?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 3]
```diff
@@ -6019,6 +6020,15 @@ public void visit(BLangCheckedExpr checkedExpr, AnalyzerData data) {
         visitCheckAndCheckPanicExpr(checkedExpr, data);
     }
 
+    @Override
+    public void visit(BLangCheckedOnFailExpr checkedOnFailExpr, AnalyzerData data) {
+        visitCheckAndCheckPanicExpr(checkedOnFailExpr.bLangCheckedExpr, data);
+        visit((BLangErrorConstructorExpr) checkedOnFailExpr.bLangErrorConstructorExpr(), data);
+        // check type against noType
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6027]: Is this comment for the above code?

**Comment by hasithaa** [comment]
This implementation is differed.

**Pull request closed without merging**